crates/services/sync/src/import/cache.rs

```diff
@@ -0,0 +1,533 @@
+                current_height,
+                end,
+                max_chunk_size,
+                chunk_size_u32,
+                end,
+            );
+        }
+        futures::stream::iter(chunks)
+    }
+
+    fn collect_cache_data(&self, range: RangeInclusive<u32>) -> Vec<(u32, CachedData)> {
+        let lock = self.0.lock();
+        lock.range(range).map(|(k, v)| (*k, v.clone())).collect()
+    }
+
+    fn handle_current_chunk(
+        &self,
+        current_chunk: CachedDataBatch,
+        data: CachedData,
+        height: u32,
+        chunks: &mut Vec<CachedDataBatch>,
+    ) -> CachedDataBatch {
+        match (current_chunk, data) {
+            (CachedDataBatch::None(_), CachedData::Header(data)) => {
+                CachedDataBatch::Headers(Batch::new(
+                    None,
+                    height..height.saturating_add(1),
+                    vec![data],
+                ))
+            }
+            (CachedDataBatch::None(_), CachedData::Block(data)) => {
+                CachedDataBatch::Blocks(Batch::new(
+                    None,
+                    height..height.saturating_add(1),
+                    vec![data],
+                ))
+            }
+            (CachedDataBatch::Headers(mut batch), CachedData::Header(data)) => {
+                batch.range = batch.range.start..height.saturating_add(1);
+                batch.results.push(data);
+                CachedDataBatch::Headers(batch)
+            }
+            (CachedDataBatch::Blocks(mut batch), CachedData::Block(data)) => {
+                batch.range = batch.range.start..height.saturating_add(1);
+                batch.results.push(data);
+                CachedDataBatch::Blocks(batch)
+            }
+            (CachedDataBatch::Headers(headers_batch), CachedData::Block(block)) => {
+                chunks.push(CachedDataBatch::Headers(headers_batch));
+                CachedDataBatch::Blocks(Batch::new(
+                    None,
+                    height..height.saturating_add(1),
+                    vec![block],
+                ))
+            }
+            (CachedDataBatch::Blocks(blocks_batch), CachedData::Header(header)) => {
+                chunks.push(CachedDataBatch::Blocks(blocks_batch));
+                CachedDataBatch::Headers(Batch::new(
+                    None,
+                    height..height.saturating_add(1),
+                    vec![header],
+                ))
+            }
+        }
+    }
+
+    fn check_chunk_limit(
+        &self,
+        current_chunk: CachedDataBatch,
+        max_chunk_size: usize,
+        chunks: &mut Vec<CachedDataBatch>,
+        current_height: u32,
+    ) -> CachedDataBatch {
+        match current_chunk {
+            CachedDataBatch::Headers(batch) => {
+                if batch.results.len() >= max_chunk_size {
+                    chunks.push(CachedDataBatch::Headers(batch));
+                    CachedDataBatch::None(current_height..current_height)
```

Comment:
I don't see why we want to return `None` instead of new `Headers` with remaining elements.

I see that it was extracted from the loop and there it makes sense, because `None` is a default value to start the next iteration of the loop. But here, it looks strange.

I think if we had function with name `truncate_chunk` and did something like `current_chunk = truncate_chunk(current_chunk, &mut chunks)`, it would be simpler to understand=)

Reply:
I refactored this to split only when the chunk is inserted in the accumulator and it really simplify the whole code. I added some comments also.